src/pybind/mgr/mds_autoscaler/module.py

```diff
@@ -0,0 +1,152 @@
+            self._orchestrator_wait([completion])
+            orchestrator.raise_if_exception(completion)
+            result = completion.result
+        except Exception as e:
+            self.log.exception("{}: exception while fetching service info"
+                               .format(e))
+            pass
+        return result
+
+    def get_daemons(self, fs_name: str) -> Optional[List[orchestrator.DaemonDescription]]:
+        result = None
+        try:
+            service = "mds.{}".format(fs_name)
+            completion = self.list_daemons(service_name=service)
+            self._orchestrator_wait([completion])
+            orchestrator.raise_if_exception(completion)
+            result = completion.result
+        except Exception as e:
+            self.log.exception("{}: exception while fetching service info"
+                               .format(e))
+            pass
+        return result
+
+    def update_daemon_count(self, fs_name: str, abscount: int) -> Optional[ServiceSpec]:
+        svclist = self.get_service(fs_name)
+        if svclist is None or len(svclist) == 0:
+            self.log.warn("Failed to fetch MDS service details for fs '{}'"
+                          .format(fs_name))
+            return None
+
+        svc = svclist[0]
+
+        # no need to update placement count if the current placement count is
+        # the required count
+        if svc.spec.placement.count == abscount:
+            return None
+
+        ps = copy.deepcopy(svc.spec.placement)
+        ps.count = abscount
+        newspec = ServiceSpec(service_type=svc.spec.service_type,
+                              service_id=svc.spec.service_id,
+                              placement=ps)
+        return newspec
+
+    def get_required_standby_count(self, fs_name: str) -> Optional[int]:
+        total = 0
+        assert self.fs_map is not None
+        for fs in self.fs_map['filesystems']:
+            if fs['mdsmap']['fs_name'] == fs_name:
+                return fs['mdsmap'].get('standby_count_wanted')
+        return total
+
+    def get_current_standby_count(self, fs_name: str, daemons: Optional[List[orchestrator.DaemonDescription]] = None) -> int:
+        # standbys are not grouped by filesystems in fs_map
+        # available = standby_replay + standby_active
+        assert self.fs_map is not None
+        total = 0
+        if daemons is None:
+            return total
+        daemon_names = {
+            d.name() for d in daemons
+        }
+        for sb in self.fs_map['standbys']:
+            full_name = "mds.{}".format(sb['name'])
+            if full_name in daemon_names:
+                total += 1
+        return total
+
+    def get_current_active_count(self, fs_name: str):
+        assert self.fs_map is not None
+        for fsys in self.fs_map['filesystems']:
+            if fsys.get('mdsmap').get('fs_name') == fs_name:
+                return len(fsys.get('mdsmap').get('in'))
```

Comment:
`in` is the set of ranks. Some may be "failed". You want `up`.

```suggestion
                return len(fsys.get('mdsmap').get('up'))
```

Reply:
ack